fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownTopNWithOuterJoin.java

```diff
@@ -0,0 +1,100 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.transformation;
+
+import com.google.common.collect.Lists;
+import com.starrocks.analysis.JoinOperator;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.base.Ordering;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalJoinOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalTopNOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class PushDownTopNWithOuterJoin extends TransformationRule {
+    public PushDownTopNWithOuterJoin() {
+        super(RuleType.TF_PUSH_DOWN_TOPN_OUTER_JOIN,
+                Pattern.create(OperatorType.LOGICAL_TOPN).addChildren(Pattern.create(OperatorType.LOGICAL_JOIN)));
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalTopNOperator topn = (LogicalTopNOperator) input.getOp();
+
+        OptExpression childExpr = input.inputAt(0);
+        LogicalJoinOperator joinOperator = childExpr.getOp().cast();
+        JoinOperator joinType = joinOperator.getJoinType();
+
+        if (!joinType.isLeftOuterJoin() && !joinType.isRightOuterJoin()) {
```

Comment:
outer join is also suit this optimization

Reply:
we will consider full outer join later.